src/main/scala/fi/oph/koski/validation/KoskiValidator.scala

```diff
@@ -83,51 +85,57 @@ class KoskiValidator(
     }
   }
 
-  private def validateOpiskeluoikeus(opiskeluoikeus: Opiskeluoikeus, henkilö: Option[Henkilö])(implicit user: KoskiSpecificSession, accessType: AccessType.Value): Either[HttpStatus, Opiskeluoikeus] = opiskeluoikeus match {
-    case opiskeluoikeus: KoskeenTallennettavaOpiskeluoikeus =>
-      updateFields(opiskeluoikeus).right.flatMap { opiskeluoikeus =>
-        (validateAccess(opiskeluoikeus)
-          .onSuccess {
-            validateLähdejärjestelmä(opiskeluoikeus)
-          }
-          .onSuccess {
-            validatePäätasonSuoritukset(opiskeluoikeus)
+  private def validateOpiskeluoikeus(opiskeluoikeus: Opiskeluoikeus, henkilö: Option[Henkilö])(implicit user: KoskiSpecificSession, accessType: AccessType.Value): Either[HttpStatus, Opiskeluoikeus] = {
+    val tallennettuOpiskeluoikeus = opiskeluoikeus.oid.flatMap(opiskeluoikeudenOid =>
```

Comment:
Hyvä, että tämä on siirretty tähän, eikä duplikoitu.

Itseäni mietityttää tässä yleisesti transaktioturvallisuus: jos datan muutoksia varten pitää tutkia vanhaa dataa, niin mielestäni oletusarvoisesti koko operaatio pitäisi tehdä samassa tietokantatransaktiossa. Ja jos/kun näin ei ole tehty, niin voisi ehkä mainita, että miksi näin ei tarvitse tehdä: luultavasti tämä uusi ja edelisestkin vanhaan sisältöön vertaavat validaatiosäännöt on sellaisia, että välissä tapahtuvat tallennukset eivät voi opiskeluoikeuden muutosta hajottaa validoitumattomaksi, vaikka transaktion väliin joku tallennus ehtisikin...

Reply:
Hmm niin, näitä transaktionaalisuutta rikkovia kohtia on muutama muukin. Taidetaan muunmuassa joissain tapauksissa sallia vain yhden päätason suorituksen siirto ja sitten loput täytetään automaattisesti.

Laitanpa siitä maininnan koodiin, tosin käytännössä lienee niin, että näissä tapauksissa tätä transaktionaalisuuden rikkomista ei voi oikein enää korjata.

Reply:
Hmm niin, Koskessa on muutama muukin kohta, jossa rikotaan transaktionaalisuusperiaatetta; esimerkiksi joissain tapauksissa voidaan siirtää vain yksi päätason suoritus ja muut suoritukset täytetään automaattisesti olemassaolevan datan pohjalta.

Sitenhän Koski pyrkii rajoittamaan virhetilanteita, että versionumeron siirretyllä opiskeluoikeudella pitää olla sama kuin edellisellä tallennetulla opiskeluoikeudella ja tämä tarkistetaan ihan pohjalla tietokataan tallentamisen funktioketjussa.

Mutta kun nyt nopeasti katsoin tuota koodia, niin ei sekään vaikuta transaktionaaliselta; mielestäni ensin haetaan vanha opiskeluoikeus ja sitten eri transaktiossa vasta päivitetään se, kun ollaan ensin scala-koodissa tarkastettu versionumeroiden mätsäys.

Pistän lyhyen kommentin kuitenkin tuosta mukaan koodiin.